### PR TITLE
draft for telephone naming automation

### PR DESCRIPTION
**Pull request by m-hynek**
#29 

**Monuments**
 - if there is `displayPhrase` in `MonumentInfo`, show it as telephone name (**Abandoned Cabins**)
   - some **difference** in certain vanilla telephone vs landmark name (Satellite Dish Array vs **Satellite Dish**)
 - append grid coords if multiple instances of same monument (**Substation G12**)
 - when there is no `displayPhrase` in `MonumentInfo` (cave_large_sewers_hard), name will be just `telephone.GetDisplayName()`+ coord (**Telephone G12**)

**Freight Transit Line**
 - Utilizing aliases + grid coord with FTL prefix
   - **FTL Train Station G12**
   - **FTL Large Intersection F23**
   - **FTL Corner Tunnel R7**

**Underwater Labs**
 - I couldn't figure how to get "Underwater Lab" string from game, so I hardcoded it for now
 - Underwater Lab + grid coord + level index + room type + room index
   - **Underwater Lab X1 L4 Room 1**
   - **Underwater Lab X1 L1 Corridor 7**
  
I am not sure if `DoServerDestroy()` is called on `Telephone` instance when unloading MA. If not, we should add `TelephoneManager.DeregisterTelephone(phoneController)` to `PreEntityKill()`

I feel like code needs some improvements, but im pretty happy with functional result

Edit: Added dynamic monuments support

### CODE REVIEW

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:4456]: While testing, I realized my current map has 2 harbors (`harbor_1` and `harbor_2`) which both have vanilla telephones. The names are `Harbor AE7` and `Harbor N31`. It's worth noting that the plugin's behavior is different; when spawning a telephone with the plugin, it's auto generated name is `Harbor`. From the standpoint of the plugin, that's correct because the telephone was only spawned at one of the harbors, because the two harbors have completely different layouts.

My concern is that if somebody wanted to do something similar, such as add telephones to multiple different harbors or multiple different fishing villages (there are two small variants both called "Fishing Village"), that players wouldn't be able to differentiate them due to not having the grid coordinates.

To solve that, I think the best option is to hard code specific monument prefabs to always display the grid coordinate, such as for harbors, fishing villages and bunker entrances. My hunch is that vanilla is doing something similar, simply setting `AppendGridToName` property to true for harbor prefabs and others, meaning the grid coordinate probably gets appended to phones at those monuments regardless of whether a given map actually has multiple instances of that monument.

**Review by WheteThunger — CHANGES_REQUESTED** [review]
This is a great start! Thanks for taking the time to implement and test this, and thank you for your patience as we work through the feedback and remaining unknowns.

One thing we should think about is pre-emptively moving some of this logic into a helper method, to prepare for the case where server owners can rename individual telephones. I realize that since what you have implemented here is pretty comprehensive, we might have eliminated any need for server owners to implement custom names, but that remains to be verified.

If we allow manually naming telephones someday, we will probably want to share/reuse the logic to append grid coordinates, and will need to implement "update" logic in the `HandleChanges()` method to rename existing phones that match the one the player is aiming at.

Also, I realize that I have presented you with a lot of things to think about in this PR and in our other discussions, so I just want to say thank you for continuing the dialogue. The reality is that there are a lot of problems to solve, and it means a lot to me that you are sticking around and helping me figure out how best to solve them.

**Comment by m-hynek** [comment]
Thank you for detailed review! 

Working (and hopefuly not bothering you much) on project like MA is great opportunity for me to learn and improve in C# and Rust plugin making. I was developing website applications in PHP for almost decade, so this is whole new world. Even if I have solid programming background, I have no real idea what I am doing most of the time, its all guessing (I still can't believe my 2 little plugins got approved on Umod lol). So thank you for keeping me around, I learned a lot in past weeks.

I will push changes tomorrow, hopefully.

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs:4466]: I believe 'Freight Transit Line' is little easter egg joke from Facepunch, as FTL generally stands for 'Faster than light'. I tought it would be nice to add this to phone names as prefix, instead of words like 'Metro' or 'Underground' and help spread the fun :)

Comment (m-hynek) [reply MonumentAddons.cs]: Regex is now compiled, but Im not sure where is best place for its initialization, so I added it as plugin property for now.

I added attached monument name to train stations. I can't imagine how pseudo-monuments would work, but since we know what monument is station attached to, we can let user set `SpawnOnThisStationOnly` option per entity, save imonument shortname to profile and before spawn, check if attached monument shortname matches saved one. This would allow having both 'global' and 'station-specific' entities at same time.

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: I like the idea of aliases for modules, but I didnt implement it yet as I need to spawn them in rustedit, take a look at them and come up with descriptive names. I will get back to this.

Room index incremented, strings refactored

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: I forgot that there are multiple versions of same monument while implementing logic for appending coordinates.

As advised, I hardcoded shortnames of multi-version prefabs into switch. Would list or something else be better approach?

**Review by WheteThunger — CHANGES_REQUESTED** [review]

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: I added static `PhoneUtils` class, but method is still quite long. Further refactoring would imo require non-static class for telephone naming. Also im not sure if `DungeonEntranceMapper` belongs to helpers region, or somewhere else.

**Review by WheteThunger — CHANGES_REQUESTED** [review]

File: MonumentAddons.cs
Comment (WheteThunger) [reply MonumentAddons.cs]: I was thinking the methods in `PhoneUtils` could take a few arguments to be responsible for branching based on monument type, but not required as of now.

Comment (m-hynek) [reply MonumentAddons.cs]: I tought passing `PluginInstance` as argument wouldn't be ideal, `PluginInstance.IsMonumentUnique` and `PluginInstance._dungeonEntranceMapper` cannot be accessed in static class. `SplitCamelCase` can be static so I will refactor some more logic into PhoneUtils. If passing `PluginInstance` as method argument is tolerable, i can move everyhing there. Or am I missng something?

Comment (WheteThunger) [reply MonumentAddons.cs]: Passing PluginInstance is okay. We can always refactor more later. We might want to eventually add some object that wraps the plugin instance to provide the monument utilities.

Comment (m-hynek) [reply MonumentAddons.cs]: Whole telephone logic is now inside `PhoneUtils`.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3684]: This code is unused.

**Review by WheteThunger — CHANGES_REQUESTED** [review]
While there are some improvements that could be made, I'm inclined to approve this after you remove the unused method, so you can focus your efforts on other features.

File: MonumentAddons.cs
Comment (WheteThunger) [reply MonumentAddons.cs]: Normally I would use an array.

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: If there are potential improvements, I dont mind to implement them. I can apply what I learn here in future PRs, so as long as you are not bothered with code reviews, dont mind me rewriting it. I know this code is not ideal, I hacked togheter some idea and it is basically same code moved elsewhere, so yeah, all insights are welcomed and very helpful.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3678]: Don't expose the adapter to this method. Instead, pass the monument, and the position of the adapter.

**Review by WheteThunger — CHANGES_REQUESTED** [review]
Since you mentioned you are open to more feedback, I'll continue adding feedback, but sorry if there are delays since I have been busy lately.

**Review by WheteThunger — CHANGES_REQUESTED** [review]

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: For now I removed adapter from helper parameters. Where I should instantiate `MonumentHelper` class? I was thinking to put it where `DungeonEntranceMapper ` currently is (plugin property), and then pass it to `PhoneUtils` in `SingleEntityAdapter` but it looks to me like almost same thing. Alternatively new instance for every `PhoneUtils.NameTelephone` call, but it seems to be

```C#
var telephone = Entity as Telephone;
if (telephone != null && telephone.prefabID == 1009655496)
{
    PhoneUtils.NameTelephone(telephone, Monument, Position, PluginInstance._monumentHelper);
}
```

Comment (WheteThunger) [reply MonumentAddons.cs]: `MonumentHelper` instance should be a field of the plugin, similar to the `DungeonEntranceMapper` instance. Could even merge the `DungeonEntranceMapper` code into the `MonumentHelper` class to reduce the number of different helpers needed. Definitely don't create more instances since we want to avoid heap allocations that will be quickly dereferenced (a.k.a. "garbage"). Above looks fine.

Comment (m-hynek) [reply MonumentAddons.cs]: `DungeonEntranceMapper` merged into `MonumentHelper`, removed plugin instance from `Phone utils`

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:148]: Replace `_pluginInstance` with `this`.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3044]: Can remove `_pluginInstance.` since this is already inside an instance method of the plugin class.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3065]: Can remove `_pluginInstance.` since this is already inside an instance method of the plugin class.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3612]: Should be named PascalCase since it's public. Arguably `DungeonEntranceMapper` doesn't need to exist as a distinct class, but I'll leave that up to you for now (we're not following OOP very strictly).

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3620]: I suggest inlining this construction to where the field is defined since it takes no parameters.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3839]: An array is fine, but we shouldn't construct it every time this method is called. Instead, make it a `static readonly` field of the class, and give it a more descriptive name.

**Review by WheteThunger — CHANGES_REQUESTED** [review]

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: removed argument from `OnServerInitialized` and resolving dependency in plugin constructor

File: MonumentAddons.cs
Comment (m-hynek) [reply MonumentAddons.cs]: Since its just 2 methods, `DungeonEntranceMapper` refactored into `MonumentHelper` class directly

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3763]: This method appears to be unused.

File: MonumentAddons.cs
Comment (WheteThunger) [inline MonumentAddons.cs:3765]: Methods that are only called within this class should be marked private.

**Review by WheteThunger — CHANGES_REQUESTED** [review]
This is very close, just a few minor changes remaining. Thanks for your patience!

**Comment by m-hynek** [comment]
Nono, thanks for your patience!

**Pull request merged**
